core/src/main/java/org/elasticsearch/client/transport/TransportClientNodesService.java

```diff
@@ -211,7 +211,13 @@ public TransportClientNodesService removeTransportAddress(TransportAddress trans
     }
 
     public <Response> void execute(NodeListenerCallback<Response> callback, ActionListener<Response> listener) {
-        List<DiscoveryNode> nodes = this.nodes;
+        List<DiscoveryNode> nodes;
+        synchronized (mutex) {
```

Comment:
@nik9000 That's not as simple as it sounds. For example, sampling modifies the list of nodes (state). During the process of sampling, any additional nodes are connected to. When closing, we need to ensure that we disconnect from all nodes connected to. That's one reason that there is a mutex to synchronize these actions, so that close is *always* guaranteed to disconnect from every node that is connected to, and no additional nodes can be connected to after we have closed. If we encapsulate all of the mutable state into a single reference that we replace when we change state, there's no guarantee in a race between adding nodes and closing the client.